### PR TITLE
Make `AttributesInterface` immutable

### PR DESCRIPTION
**Pull request by Nevay**
Makes `AttributesInterface` immutable by introducing `AttributesBuilderInterface` as mutable builder.

Related:
https://github.com/open-telemetry/opentelemetry-php/pull/521#issuecomment-1003330123
https://github.com/open-telemetry/opentelemetry-php/pull/520

### CODE REVIEW

File: src/SDK/AttributesBuilder.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        return isset($this->attributes[$offset]);
+    }
+
+    public function offsetGet($offset)
+    {
+        return $this->attributes[$offset] ?? null;
+    }
+
+    public function offsetSet($offset, $value): void
```
Comment (lalex) [inline src/SDK/AttributesBuilder.php:95]: An `$offset` which is an Attribute key "must be a non-null and non-empty string" ([spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/common/common.md#attributes)). It's better to add type hint `string`.

File: src/SDK/Trace/Span.php [hunk 2]
```diff
@@ -361,9 +360,9 @@ public function toSpanData(): SpanDataInterface
             $this->name,
             $this->links,
             $this->events,
-            $this->getImmutableAttributes(),
-            (null === $this->attributes) ? 0 : $this->attributes->getTotalAddedValues(),
-            $this->totalRecordedEvents,
+            $this->attributes->build(),
+            max(0, $this->droppedEventsCount),
```
Comment (lalex) [inline src/SDK/Trace/Span.php:364]: I've spent a while to understand why the count could be negative number. Very interesting approach to store the limit value.

File: src/SDK/Trace/SpanBuilder.php [hunk 3]
```diff
@@ -36,24 +34,39 @@ final class SpanBuilder implements API\SpanBuilderInterface
...
         InstrumentationLibrary $instrumentationLibrary,
-        TracerSharedState $tracerSharedState,
-        SpanLimits $spanLimits
+        ResourceInfo $resource,
+        SamplerInterface $sampler,
+        SpanProcessorInterface $spanProcessor,
+        IdGeneratorInterface $idGenerator,
+        AttributesBuilderInterface $attributes,
+        AttributesFactoryInterface $linkAttributes,
+        AttributesFactoryInterface $eventAttributes,
```
Comment (lalex) [inline src/SDK/Trace/SpanBuilder.php:55]: It's not clear when Attributes are the Builder and when they are Factory. Probably the proper naming of the arguments and properties could help with it.

File: src/SDK/Trace/SpanLimits.php [hunk 4]
```diff
@@ -25,6 +27,30 @@ final class SpanLimits
 
     private int $attributePerLinkCountLimit;
 
+    public function getSpanAttributes(): AttributesFactoryInterface
```
Comment (lalex) [inline src/SDK/Trace/SpanLimits.php:30]: Is it correct to have a creational methods in the value-class?

File: src/SDK/AttributesBuilder.php (hunk 1 above)
Comment (brettmc) [reply src/SDK/AttributesBuilder.php:95]: Also, needs to be compatible with ArrayAccess signature for 8.1+, see https://github.com/open-telemetry/opentelemetry-php/runs/4729027088?check_suite_focus=true#step:13:45

File: src/SDK/Trace/Event.php [hunk 5]
```diff
@@ -14,11 +13,11 @@ final class Event implements EventInterface
     private int $timestamp;
     private AttributesInterface $attributes;
 
-    public function __construct(string $name, int $timestamp, AttributesInterface $attributes = null)
+    public function __construct(string $name, int $timestamp, AttributesInterface $attributes)
```
Comment (brettmc) [inline src/SDK/Trace/Event.php:16]: The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events) says "zero or more attributes" for an event, so I think they should remain optional?

File: src/SDK/Trace/Sampler/TraceIdRatioBasedSampler.php [hunk 6]
```diff
@@ -60,7 +60,7 @@ public function shouldSample(
         $traceIdCondition = $lowerOrderBytes < round($this->probability * $traceIdLimit);
         $decision = $traceIdCondition ? SamplingResult::RECORD_AND_SAMPLE : SamplingResult::DROP;
 
-        return new SamplingResult($decision, $attributes, $traceState);
+        return new SamplingResult($decision, [], $traceState);
```
Comment (brettmc) [inline src/SDK/Trace/Sampler/TraceIdRatioBasedSampler.php:63]: why not passing `$attributes` through?

File: src/SDK/Trace/Span.php [hunk 7]
```diff
@@ -66,10 +63,10 @@ final class Span extends API\AbstractSpan implements ReadWriteSpanInterface
     /** @var list<EventInterface> */
     private array $events = [];
 
-    private ?AttributesInterface $attributes;
-    private int $totalRecordedEvents = 0;
+    private AttributesBuilderInterface $attributes;
```
Comment (brettmc) [inline src/SDK/Trace/Span.php:66]: I don't get this, the variable might be better named `$attributesBuilder` ?

File: src/SDK/Trace/Sampler/TraceIdRatioBasedSampler.php (hunk 6 above)
Comment (Nevay) [reply src/SDK/Trace/Sampler/TraceIdRatioBasedSampler.php:63]: The returned attributes are additional attributes, ["A set of span Attributes that will also be added to the `Span`"](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/trace/sdk.md#shouldsample).

File: src/SDK/Trace/Event.php (hunk 5 above)
Comment (Nevay) [reply src/SDK/Trace/Event.php:16]: Callers can provide an empty `AttributesInterface` instance; allowing `null` to represent empty attributes felt unecessary as the only caller (`Span`) always passes non-null attributes.

**Comment by bobstrecansky** [comment]
Hey @Nevay -  are you still planning on updating this PR?

File: src/SDK/Trace/Span.php (hunk 2 above)
Comment (Nevay) [reply src/SDK/Trace/Span.php:365]: Should I revert this change or is adding a comment to the `droppedLinksCount` and `droppedEventsCount` properties sufficient?

**Comment by bobstrecansky** [comment]
Closing this due to inactivity; please feel free to reopen if you'd like to!

**Pull request closed without merging**